### ISSUE TITLE
Auth.send - TypeError: Cannot read property 'auth_token' of undefined

### ISSUE THREAD

**Issue by imjonsnooow** [heartsentwined/ember-auth#56]
While trying to use Auth.send, we get the TypeError included in the issue title. We need the auth_token to end up at the server, but this error prevents that. I can't tell if it's a configuration problem, or if there's a problem in the underlying code.

Here's my Auth configuration:

```
Auth = Ember.Auth.create({
signInEndPoint: '/users/sign_in',
signOutEndPoint: '/users/sign_out',
tokenKey: 'auth_token',
tokenIdKey: 'user_id',
userModel: 'FlexiconEmber.User',
urlAuthenticationRouteScope: 'both',
modules: ['emberData', 'rememberable'],
rememberable: {
    tokenKey: 'remember_token',
    period: 7,
    autoRecall: true
}
})
Auth.currentUser = null;
Auth.on('signInComplete', function() {
Auth.set("currentUser", Auth.get('user'));
});
Auth.on('signOutComplete', function() {
Auth.set("currentUser", null);
});
auth = Auth;
```

The line in ember-auth.js that we're crashing on:

```
get$(opts, 'data')[get$(get$(this, 'auth'), 'tokenKey')] || (get$(opts, 'data')[get$(get$(this, 'auth'), 'tokenKey')] = get$(this, 'authToken'));
```

What am I doing wrong?

**Comment by heartsentwined** [comment]
@scaswell1 would you mind upgrading to the latest version (v`6.0.3`)? The above style shows v`4.x` or below, and it's pretty much spaghetti code back then. The [upgrade guide](https://github.com/heartsentwined/ember-auth/blob/master/CHANGELOG.md#upgrade-guide) can help, and the official [docs](http://ember-auth.herokuapp.com) is actually a code-generator that will generate the `ember-auth` code you need for your particular use case.

Let me know if you cannot upgrade for some reason, but it'll take longer to debug, because I no longer have a sandbox for the old version and I'll need to get everything setup again.

**Comment by imjonsnooow** [comment]
@heartsentwined my Rails `bundle show` says I'm running `ember-auth-rails (4.0.1)` and `ember-auth-source (6.0.3)`.

I think I am using the latest version, but possibly in the incorrect style -- I just recently updated to this new version. Most of what I'm seeing in my code seems to me in keeping with the information listed on the official docs.

**Comment by heartsentwined** [comment]
1. The docs suggest that you scope your app's `Auth` within the app namespace (`FlexiconEmber.Auth`) to ensure that it is available app-wide. You can use an unscoped `Auth`, just make sure that it is available (not `var`-ed, or make it on `window.Auth`, etc).
2. `urlAuthenticationRouteScope` is gone. (`both` is now the default; having that line there won't hurt, it just does nothing.)
3. But I see you have not enabled the `urlAuthenticatable` module. Do you still need its functionality?
4. What does the final `auth = Auth` do?
5. I see that you are reviving the old API `Auth.currentUser` manually. There's nothing wrong with that, but be mindful that `Auth.get('user')` is loaded _asynchronously_ (it was also loaded async before). Don't rely on it to be available immediately.
6. Make sure that `ember-auth` gets initialized _before_ `ember` does its bootstrap. Usually you would also need to do this _after_ initializing the app itself, but since you are not scoping `Auth` under it, so it doesn't apply for you. Rails example:

`application.coffee`:

``` coffeescript
# ...
#= require ember
#= require ember-auth
#= require_self
# ...
#= require_tree .
window.App = Ember.Application.create()
App.Auth = Ember.Auth.create() # or just Auth, instead of App.Auth, as you like
```

Does this help? Otherwise please post more situational information, viz. on what conditions you are using the `Auth.send()`, perhaps some surrounding code, or info like when did the user signed in (if any), whether it's on some specific route, etc. Another useful information would be a sample of your server's (successful) sign in response, just to make sure that the response can be decoded correctly.

**Comment by imjonsnooow** [comment]
1. I have gone through and scoped `Auth` within our `FlexiconEmber` namespace for good measure.
2. As `urlAuthenticationRouteScope` defaults, I have removed it from the Auth create call.
3. I believe we may, and I have added `urlAuthenticable` back into the config
4. I noticed in the offending line in my original post that `get$(get$(this, 'auth')` was undefined unless I had `auth` initialized, so I was testing to see if that resolved the issue (sadly, it did not.)
5. Noted, we have it working reliably! :)
6. What follows is now my application.js file:
   
   ```
   //= require jquery
   //= require jquery_ujs
   //= require twitter/bootstrap
   //= require handlebars
   //= require ember
   //= require ember-data
   //= require ember-auth
   //= require_self
   //= require flexicon_ember
   //= require_tree .
   //
   //Ember.LOG_BINDINGS = true;
   FlexiconEmber = Ember.Application.create({});
   FlexiconEmber.Auth = Ember.Auth.create({
       signInEndPoint: '/users/sign_in',
       signOutEndPoint: '/users/sign_out',
       tokenKey: 'auth_token',
       tokenIdKey: 'user_id',
       userModel: 'FlexiconEmber.User',
       modules: ['rememberable', 'emberData', 'urlAuthenticatable'],
       rememberable: {
       tokenKey: 'remember_token',
           period: 7,
           autoRecall: true
        }
   });
   FlexiconEmber.Auth.currentUser = null;
   FlexiconEmber.Auth.on('signInComplete', function() {
   FlexiconEmber.Auth.set("currentUser", FlexiconEmber.Auth.get('user'));
   });
   FlexiconEmber.Auth.on('signOutComplete', function() {
       FlexiconEmber.Auth.set("currentUser", null);
   });
   ```

The use case for our `Auth.send` is: we have models which can be voted on (upvote/downvote) using UI controls. The user has to be able to do this quickly and easily - they do so by clicking buttons that only appear if there is a user currently logged in.

The Auth.send currently looks as follows (maybe I'm doing that wrong?):

```
   // scope is within an Ember ObjectController
   FlexiconEmber.Auth.send('upvote/' + ctrl.get('model').get('id') + '.json', function(data) {
        if (data && !data.already_voted) {
            ctrl.get('model').reload();
        } else {
            FlexiconEmber.set("error", "You've already voted for this item!")
            $('#myModal').modal('show');
        }
    });
```

**Comment by heartsentwined** [comment]
The `App.Auth.send()` method expects polymorphic parameters depending on the underlying `requestAdapter` that was chosen. This defaults to `jquery`, and the `jquery` adapter delegates to [`jQuery.ajax()`](http://api.jquery.com/jQuery.ajax/#jQuery-ajax-settings) (after adding auth data). In particular, it uses [`jquery.ajax(settings)`](http://api.jquery.com/jQuery.ajax/#jQuery-ajax-settings), where `settings` is an object of options ([read the docs](http://api.jquery.com/jQuery.ajax/#jQuery-ajax-settings)).

The first part 

``` javascript
'upvote/' + ctrl.get('model').get('id') + '.json'
```

is obviously an URL, so I'd put it this way

``` javascript
FlexiconEmber.Auth.send({
    url: 'upvote/' + ctrl.get('model').get('id') + '.json' 
});
```

I'm not clear what this code is supposed to do:

``` javascript
function(data) {
    if (data && !data.already_voted) {
        ctrl.get('model').reload();
    } else {
        FlexiconEmber.set("error", "You've already voted for this item!")
        $('#myModal').modal('show');
    }
}
```

It looks like a callback on success, so I'd do

``` javascript
FlexiconEmber.Auth.send({
    url: 'upvote/' + ctrl.get('model').get('id') + '.json' 
}).done(function(data) {
    if (data && !data.already_voted) {
        ctrl.get('model').reload();
    } else {
        FlexiconEmber.set("error", "You've already voted for this item!")
        $('#myModal').modal('show');
    }
});
```

**Comment by imjonsnooow** [comment]
That makes a lot of sense. After going through this, I started to suspect I was utilizing the `send` method wrong.
However, this approach attempted to access the following url: 

```
 upvote/68.json?{%22auth_token%22:%227zr5PJffhbFzTUivmy5R%22}
```

It seems the auth_token is being appended incorrectly. Is there a setting to manage this?

**Comment by heartsentwined** [comment]
One more note: `urlAuthenticatable` needs further configuration. Please refer to the **Pre-authentication by URL query params** section of the [docs](http://ember-auth.herokuapp.com/docs) for details.

**Comment by heartsentwined** [comment]
Let me give a try in my sandbox, and I'll get back to you once I nailed down the problem.

**Comment by heartsentwined** [comment]
One more thing: what is the expected HTTP verb for `upvote/68.json`? `POST`?

**Comment by imjonsnooow** [comment]
Thanks - I'll wait to hear back.
Also, they're set up as `GET` routes, but that could be changed if necessary.

**Comment by heartsentwined** [comment]
@scaswell1 sorry for the wait; it's a bug, and fixed on `6.0.5` now.
